### PR TITLE
Use StoppableWorkers in the GPS DataReaders

### PR DESCRIPTION
**Pull request by penguinland**
I recommend reviewing this while suppressing whitespace changes; otherwise the diff is way bigger than it should be (60ish lines of code merely became less indented).

Changes include:
- `start()` in both the `PmtkI2cDataReader` and `SerialDataReader` classes has been renamed `backgroundWorker`, and no longer spawns a background goroutine (it's just the implementation of the code that should run in the background). 
- `gpsutils.NewI2cDataReader` takes in a context. Should `gpsutils.NewSerialDataReader` take in a context it doesn't use, just to keep consistency? I'm unsure.
- The background goroutines in both data readers check their context while trying to put data into the output channel. This greatly simplifies the implementation of `Close()`, because you don't need to read from the output channel to trigger the background worker shutting down.

Tried on John-Pi: the NMEA component seems to work fine with both serial and I2C connections, and the RTK component seems to work fine with both VRS and non-VRS casters, with both serial and I2C connections.

### CODE REVIEW

File: components/movementsensor/gpsutils/i2c_data_reader.go
Comment (JohnN193) [inline components/movementsensor/gpsutils/i2c_data_reader.go:109]: [giganit/can ignore] `ctx` instead of `cancelCtx`

File: components/movementsensor/gpsutils/i2c_data_reader.go
Comment (JohnN193) [inline components/movementsensor/gpsutils/i2c_data_reader.go:128]: [giganit/can ignore] ctx instead of cancelCtx

**Review by JohnN193 — APPROVED** [review]
LGTM, only thought is if you want to change a variable name used in some functions, but otherwise makes sense to me

**Comment by JohnN193** [comment]
> gpsutils.NewI2cDataReader takes in a context. Should gpsutils.NewSerialDataReader take in a context it doesn't use, just to keep consistency? I'm unsure.

personally I don't think we should pass in the variable that gets unused, but we also made a similar decision in other parts of the gps code to keep things consistent. So I guess my vote is for consistency

**Review by susmitaSanyal — APPROVED** [review]

**Comment by penguinland** [comment]
John, that was excellent feedback! I've made all your suggested changes, and will test again before merging again (but not during the bug bash today)

**Comment by penguinland** [comment]
Tried one last time on John-Pi: all 6 GPS devices (NMEA, RTK with a caster, RTK with a VRS) (I2C and serial interfaces) work like normal.

**Pull request merged**
